cpp/dolfinx/common/IndexMap.cpp

```diff
@@ -32,6 +32,353 @@ std::array<std::vector<int>, 2> build_src_dest(MPI_Comm comm,
+compute_submap_indices(const dolfinx::common::IndexMap& imap,
+                       std::span<const std::int32_t> indices)
+{
+  // --- Step 1 ---: Send ghost indices in `indices` to their owners
+  // and receive indices owned by this process that are in `indices`
+  // on other processes
+
+  const MPI_Comm comm = imap.comm();
+  // TODO Should these be spans?
+  const std::vector<std::int32_t>& src = imap.src();
+  const std::vector<std::int32_t>& dest = imap.dest();
+  const std::vector<std::int64_t>& ghosts = imap.ghosts();
+  const std::vector<std::int32_t>& ghost_owners = imap.owners();
+
+  // Create neighbourhood comm (ghost -> owner)
+  MPI_Comm comm0;
+  int ierr = MPI_Dist_graph_create_adjacent(
+      comm, dest.size(), dest.data(), MPI_UNWEIGHTED, src.size(), src.data(),
+      MPI_UNWEIGHTED, MPI_INFO_NULL, false, &comm0);
+  dolfinx::MPI::check_error(comm, ierr);
+
+  // Create lookup array to determine if an index is in the sub-map
+  std::vector<int> is_in_submap(imap.size_local() + imap.num_ghosts(), 0);
+  for (auto v : indices)
+    is_in_submap[v] = 1;
+
+  // Pack ghost indices that this process wants to include in the
+  // sub-map
+  std::vector<std::vector<std::int64_t>> send_data(src.size());
+  for (std::size_t i = 0; i < ghosts.size(); ++i)
+  {
+    auto it = std::lower_bound(src.begin(), src.end(), ghost_owners[i]);
+    assert(it != src.end() and *it == ghost_owners[i]);
+    int r = std::distance(src.begin(), it);
+    // Send ghost index if it is in the submap, else send -1
+    if (is_in_submap[imap.size_local() + i])
+      send_data[r].push_back(ghosts[i]);
+    else
+      send_data[r].push_back(-1);
+  }
+
+  // Count number of indices to send per dest
+  std::vector<std::int32_t> send_sizes;
+  std::transform(send_data.begin(), send_data.end(),
+                 std::back_inserter(send_sizes),
+                 [](auto& d) { return d.size(); });
+
+  // Send how many indices I ghost to each owner, and receive how many
+  // of my indices other ranks ghost
+  std::vector<std::int32_t> recv_sizes;
+  recv_sizes.resize(dest.size(), 0);
+  send_sizes.reserve(1);
+  recv_sizes.reserve(1);
+  ierr = MPI_Neighbor_alltoall(send_sizes.data(), 1, MPI_INT32_T,
+                               recv_sizes.data(), 1, MPI_INT32_T, comm0);
+  dolfinx::MPI::check_error(comm, ierr);
+
+  // Prepare displacement vectors
+  std::vector<int> send_disp, recv_disp;
+  send_disp.resize(src.size() + 1, 0);
+  recv_disp.resize(dest.size() + 1, 0);
+  std::partial_sum(send_sizes.begin(), send_sizes.end(),
+                   std::next(send_disp.begin()));
+  std::partial_sum(recv_sizes.begin(), recv_sizes.end(),
+                   std::next(recv_disp.begin()));
+
+  // Send ghost indices to owner, and receive indices owned by this process
+  // that are ghosts on others
+  std::vector<std::int64_t> send_indices, recv_indices;
+  for (auto& d : send_data)
+    send_indices.insert(send_indices.end(), d.begin(), d.end());
```

Comment:
You could make this communicator non-blocking, and pack the data while it is being sent. 
Then compute the `recv_disp` and `recv_indices` once packing of outgoing data is finished, see for instance: https://github.com/FEniCS/dolfinx/blob/6d0720e3dcdacdcd045afbb376c367a70c27fe42/cpp/dolfinx/geometry/utils.h#L733-L763